HomeKit controller refactoring (towards BLE support) [DO NOT MERGE]

Jc2k opened the pull request:
## Description:

Note: This PR builds on #20325 which is a bug fix change. I'm opening this PR in parallel as anticipate some discussion around the changes.

This pull request lays down some foundational work for supporting HomeKit BLE devices. There will be more of this when the homekit library ships BLE support (branch [here](https://github.com/jlusiardi/homekit_python/tree/add_bluetooth), but I want to keep the delta i'm running small in the meantime (easier to work on BLE AND support HA users if the delta is small).

There are a couple of issues with the current component that become a problem with BLE:

 * BLE doesn't have an equivalent to the `list_accessories_and_characteristics()` endpoint so we have to emulate it. It's slooooow. A normal homekit controller uses the `get_characteristics()` API and only polls the characteristics its tracking.

 * BLE devices are often battery powered and optimized for low interaction with the controller (e.g. favouring events). To save batteries (and polar bears) we want to interact with devices less (and eventually switch to events rather than polling).

Put another way its silly to waste time and battery asking a battery powered Eve Motion what its serial number is once a minute. We should stop doing that.

As a first step to support these goals this branch i've switched to calling `get_characteristics` for only the characteristic types we are interested in. Additionally, theres now a distinct `setup` and `update` phase. The setup phase runs at start up and configures supported features based on the discovery data. There is no need to wait for the first successful `update` to know that a light supports brightness control (which is what happens now).

There is now a per characteristic callback for decoding a value from the API into entity state. Ultimately this will be called directly by an event from the HomeKit API.

The best part of this change is of course that is accompanied by the first set of tests for homekit_controller.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Jc2k committed:
Refactor homekit_controller entities to reduce amount of data they pull

The current method polls list_accessories_and_characteristics. This fetches
everything, including values and schema metadata, for all devices on a
bridge. The limitations of this are:

 * The entity is not fully configured until the first successful poll
   even though we know enough to configure it immediately.

 * On the BLE backend, this is method very slow as it has to be
   emulated through multiple get_characteristic calls.

 * Entities have to fetch unrelated data when polling (all
   entities on bridge are included)

 * Entities have to fetch attributes they dont use

The new method is to declare up front which characteristics are relevant to
an entity and to poll them explicitly with the get_characteristics method,
We also do as much entity setup as we can when an entity is created.

This doesn't change functionality - behaviour should be as before but now all
polling is targetting - it only fetches characteristics that are relevant.

This is a precursor for bluetooth support and also for events support.

Jc2k commented:
Don't know why cla-bot has got stuck by I have signed it (see my [other PR](https://github.com/home-assistant/home-assistant/pull/20325), where it didn't get stuck).

Jc2k commented:
Rebased to fix conflict, which has also unstuck cla-bot :)

MartinHjelmare commented on tests/components/alarm_control_panel/test_homekit_controller.py:
```diff
@@ -0,0 +1,142 @@
+
+def test_get_characteristic_types():
+    """Test the entity uses 1 or more characteristics."""
+    entity = mock_alarm_control_panel()
```
Please don't test the entity in isolation. Set up the homekit controller component, via `async_setup_component`, let it set up the correct platform, let that add an entity. Then use the core to test the entity, eg by calling a service and asserting the result on the library methods. If it's a sensor, assert the state of the entity in the core state machine.

Make sure to patch and mock appropriately to avoid I/O.

MartinHjelmare replied:
I usually recommend the deconz test suite.

We usually put common things in common.py under the component in question in the tests directory.

When using pytest fixtures, common fixtures go in conftest.py under the component in question. Pytest fixtures are awesome to use when writing tests that need common functionality.

Consider moving the platforms under the homekit controller component package. That's our new way of structuring components with platforms.

Jc2k replied:
Plan! One last question - is there a platform I can look at that does this already?

MartinHjelmare replied:
Deconz. :smile:

Jc2k replied:
Great. So i'm trying to figure out how HA 'finds' light.py/cover.py in this new structure. Does `setup_platform` still work? Deconz, hue, etc use `setup_entry` instead.

Jc2k replied:
In my testing it looks like just `git mv homeassistant/components/light/homekit_controller.py homeassistant/components/homekit_controller/light.py` is enough!

MartinHjelmare replied:
Magic! That logic is in loader.py.

MartinHjelmare reviewed (commented):
Try to break up the PR in smaller pieces, if at all possible. Review will be faster then.

Jc2k commented:
It’s going to be difficult to break up into meaningful smaller pieces - I already did that and have the next 5 patches after this queued up in another branch 😀. I could remove the tests and add them in another branch/PR but that didn’t seem that helpful - without any test coverage it’s quite scary!

Jc2k committed:
homekit_controller tests: automatically find entity ids in tests

Some entities use dynamic ids because of the nature of the test fakes it is
hard to predict the name of the entity that will be created. This inspects the
EntityComponent of the domain to find the freshly created entity.

Jc2k committed:
homekit_controller: Tests can now define their own Service models.

All existing tests use models as defined upstream. But upstream only defines a
few service models. This adds a generic model helper for creating test
service/characteristic models.

Jc2k committed:
homekit_controller: Add cover tests

Jc2k committed:
homekit_controller: Add lock tests

Jc2k committed:
homekit_controller: Add alarm_control_panel tests

Jc2k committed:
homekit_controller: Update light tests for color_temp.

Jc2k commented:
I've rebased this on top of https://github.com/home-assistant/home-assistant/pull/20515. I'll rebased again against `dev` when its merged. Tests still pass without having to change anything, yay!

In terms of breaking it down further still - hmm. I'd prefer not to as i've tested it as a whole. But if it was too much to review then thats fine I could probably split it in 2 (i'd basically be reimplementing it to achieve this, though). 

So plan A - with tests already merged the PR is reviewable and proceeds as is.

Plan B - its still a monster (oh dear, you should see my next branch 😱) and i do the following PRs:

 * [ ] Leave `update()` mostly as is, but add the `setup` mechanism. This means that, for example, we only work out what the `on` characteristic is once at start up rather than on every poll, and that the `turn_on()` call will actually work before the first poll (or indeed, if there were no polling).

 * [ ] Refactor code remaining in `update()` as per this PR.

With the caveat that some of the split will be artifical - the setup part of the PR will captures metadata that I don't need until the 2nd PR, which you might find weird when reviewing. But safer, I think, that trying to create a perfect intermediate stage.

How should I proceed?

Jc2k commented:
Have split this as I suggested.

 * [`setup()` refactoring](https://github.com/home-assistant/home-assistant/compare/dev...Jc2k:homekit_controller_ble_groundwork_1a)
 * [`update()` refactoring](https://github.com/home-assistant/home-assistant/compare/dev...Jc2k:homekit_controller_ble_groundwork_1b)

If #20515 by the next time I look at my HA patch queue i'll rebase and open a PR for 1a. Otherwise I will open a PR anyway, and rebase after the merge.

I've kept the tests passing at every single commit. And i've gone further than I said I would in terms of breaking them up. E.g. I came up with a way to add the new approach whilst still supporting the old approach then gradually remove it over a couple of seperate commits. This means I add code (some of which I don't like) just to remove it. But it helps to show intent if you review each commit in turn.

Also bear in mind that the intention is to support change notifications in a future set of PR's, so there is skaffolding here that might seem like overhead in isolation.

MartinHjelmare commented:
#20515 is merged.

Jc2k commented:
Closing in favour of #20538 and [this branch](https://github.com/home-assistant/home-assistant/compare/dev...Jc2k:homekit_controller_ble_groundwork_1b).
